### PR TITLE
Chart for Spark operator

### PR DESCRIPTION
**Pull request by yuchaoran2011**
**What this PR does / why we need it**:

Adds a chart for the [Spark Operator](https://github.com/GoogleCloudPlatform/spark-on-k8s-operator).

### CODE REVIEW

**Comment by yuchaoran2011** [comment]
CLA now signed. Please verify. Thanks!

**Comment by yuchaoran2011** [comment]
Could a repo maintainer help review this PR? Thanks!

**Comment by yuchaoran2011** [comment]
Hi @mgoodness @viglesiasce, not sure if you are the right guys, but can you review the PR or assign it to the right people? Thanks!

File: incubator/sparkoperator/templates/spark-operator.yaml
Comment (mrow4a) [inline incubator/sparkoperator/templates/spark-operator.yaml]: I would force the operator image tag to be also version label ({{ .Values.operatorVersionLabel }}), what do you think @liyinan926 ?

Comment (liyinan926) [reply incubator/sparkoperator/templates/spark-operator.yaml]: Yes, I agree.

Comment (yuchaoran2011) [reply incubator/sparkoperator/templates/spark-operator.yaml]: Thanks. Now fixed.

**Comment by yuchaoran2011** [comment]
@mrow4a @liyinan926 I've updated this PR to work with the new release of the Spark operator, especially the webhook feature. I also fixed an issue with kubeVersion as explained [here](https://github.com/helm/helm/issues/3810). Can you review and merge? Thanks!

File: incubator/sparkoperator/README.md
Comment (liyinan926) [inline incubator/sparkoperator/README.md]: The link for `Mutating Admission Webhook` should better be https://github.com/GoogleCloudPlatform/spark-on-k8s-operator/blob/master/docs/quick-start-guide.md#using-the-mutating-admission-webhook.

File: incubator/sparkoperator/README.md
Comment (liyinan926) [inline incubator/sparkoperator/README.md]: `to place secrets` -> `to create the server certificate bundle for the Mutating Admission Webhook and upload the bundle to a secret...`

File: incubator/sparkoperator/README.md
Comment (liyinan926) [inline incubator/sparkoperator/README.md]: Missing `if` before `it does not exist`.

File: incubator/sparkoperator/templates/spark-operator-rbac.yaml
Comment (liyinan926) [inline incubator/sparkoperator/templates/spark-operator-rbac.yaml]: Not sure if this license header should be removed or replaced with one that's more appropriate for this repo.

File: incubator/sparkoperator/values.yaml
Comment (liyinan926) [inline incubator/sparkoperator/values.yaml]: Should this be `true` instead?

File: incubator/sparkoperator/templates/spark-operator-rbac.yaml
Comment (yuchaoran2011) [reply incubator/sparkoperator/templates/spark-operator-rbac.yaml]: Good point. I have no idea. Feel free to mention/@ anyone you think would know.

File: incubator/sparkoperator/values.yaml
Comment (yuchaoran2011) [reply incubator/sparkoperator/values.yaml]: I think it should be `true` too. But one concern is that when it's `true`, the chart assumes that the user has already created the `sparkoperator` namespace, as documented in the README. If the user is not careful reviewing the README, he'll get an error when installing the chart. If you don't think it's an issue to worry about, I'm fine changing it to `true`.

File: incubator/sparkoperator/templates/spark-operator-rbac.yaml
Comment (liyinan926) [reply incubator/sparkoperator/templates/spark-operator-rbac.yaml]: Let's remove the header for now.

File: incubator/sparkoperator/README.md
Comment (liyinan926) [inline incubator/sparkoperator/README.md]: Change the default value to `false`.

File: incubator/sparkoperator/values.yaml
Comment (liyinan926) [reply incubator/sparkoperator/values.yaml]: Makes sense, let's keep `false` as the default.

Comment (yuchaoran2011) [reply incubator/sparkoperator/values.yaml]: Cool. Thanks for the review. Now comments are all addressed. Feel free to merge or ping the right person to merge.

**Review by liyinan926 — APPROVED** [review]
LGTM. Thanks!

File: incubator/sparkoperator/README.md
Comment (yuchaoran2011) [inline incubator/sparkoperator/README.md]: A comment on the image `lightbend/sparkoperator:2.3.1`. It's built by running `docker build` command using the Dockerfile [here](https://github.com/GoogleCloudPlatform/spark-on-k8s-operator/pull/247/files).

File: incubator/sparkoperator/templates/spark-operator-with-metrics.yaml
Comment (mrow4a) [inline incubator/sparkoperator/templates/spark-operator-with-metrics.yaml]: What if I want to use both webhooks and metrics?

File: incubator/sparkoperator/templates/spark-operator-rbac.yaml
Comment (mrow4a) [inline incubator/sparkoperator/templates/spark-operator-rbac.yaml:35]: how is this handled if webhooks are used? Meaning the order?

File: incubator/sparkoperator/templates/spark-operator-with-metrics.yaml
Comment (yuchaoran2011) [reply incubator/sparkoperator/templates/spark-operator-with-metrics.yaml]: Yeah good point. In the README, I added a line saying the chart only supports metrics or webhooks, but not both. But that's probably not the right thing to do. What should the correct behaviors be if both are enabled?

File: incubator/sparkoperator/templates/spark-operator-rbac.yaml
Comment (yuchaoran2011) [reply incubator/sparkoperator/templates/spark-operator-rbac.yaml:35]: The line that you referred to here would only run when the webhook is not enabled (see the first line in the file). When it is enabled, the order is enforced using [hook weights](https://github.com/helm/charts/pull/6239/files#diff-b72874954241f266b2c52983308e8dbcR11). Note the -10, -5 and 0.

**Comment by mrow4a** [comment]
@yuchaoran2011 @liyinan926  it seems there is some sort of ordering in helm - https://github.com/helm/helm/blob/master/pkg/tiller/kind_sorter.go#L29

**Comment by yuchaoran2011** [comment]
@mrow4a Thanks for the pointer. That order is problematic for this chart because, for example, deployments are created before jobs. But in our case, we would like the job to run first (i.e. generate and install the secret bundle), and then create the deployment.

**Comment by yuchaoran2011** [comment]
The last commit failed the `pull-charts-e2e` tests because of the reason I mentioned above: the namespace was created as a pre-install hook because the webhook was enabled. The e2e tests deleted the last Helm release and tried to deploy a new release, only to find out that the namespace was still there. In the latest commit, I re-set `enableWebhook` to be `false` by default, and added a section called "Limitations" in the README, reminding the user to manually clean up the namespace. Note that no manual steps are required if the webhook is not enabled.

**Comment by yuchaoran2011** [comment]
Is this PR good to be merged now?

**Comment by mrow4a** [comment]
@yuchaoran2011 I will have chance to test it maybe end of next week ;/ Lets see @liyinan926 My main concerns are however:

- separate metrics and webhooks
- fact that namespace for spark operator is declared in different yaml depending on if web hook is enabled or not

**Comment by yuchaoran2011** [comment]
@mrow4a Thanks! Those are two points that I'm not too pleased with either. Feel free to suggest ways of refactoring/improvements that can address them when you think of any.

File: incubator/sparkoperator/templates/spark-operator-with-metrics.yaml
Comment (liyinan926) [reply incubator/sparkoperator/templates/spark-operator-with-metrics.yaml]: I think we should support both. We just need to make sure the service declares the ports for both. We only need one service for both. Is it possible to combine all there yamls into one?

Comment (mrow4a) [reply incubator/sparkoperator/templates/spark-operator-with-metrics.yaml]: It should be possible. Also to properly initialize namespaces. You might have a look how different projects do it  @yuchaoran2011

Comment (yuchaoran2011) [reply incubator/sparkoperator/templates/spark-operator-with-metrics.yaml]: Ok I'll see what I can do by early next week.

File: incubator/sparkoperator/templates/spark-operator.yaml
Comment (mrow4a) [inline incubator/sparkoperator/templates/spark-operator.yaml]: Just out of curiosity, what would happen if you just leave this block default for everything (with hook-weights), also for without webhooks?

Comment (yuchaoran2011) [reply incubator/sparkoperator/templates/spark-operator.yaml]: I could do that. But in that case, the user would need to manually delete the namespace when the chart release is deleted regardless of the setting of `enableWebhook`. The current template (i.e. with the `if` condition) saves the user from doing that when `enableWebhook` is `false`.

**Comment by liyinan926** [comment]
> As I mentioned previously, I could add a post-delete hook to delete the namespace, but that would still leave namespace creation in two places. Any suggestions?

I think it's fine to leave namespace creation in two places, but it's more important to make sure the namespace gets deleted. So I think we should use a post-delete hook. @mrow4a.

**Comment by yuchaoran2011** [comment]
@liyinan926 @mrow4a I've added a post-delete hook to clean up the namespace. Now the chart can be installed and deleted without involving any manual steps (besides the `helm install` and `helm delete`, of course) regardless of the webhook setting. 

But I see that the `pull-charts-e2e` test is failing with "Error: namespaces "sparkoperator" already exists" during chart install. Is the build machine keeping some bad state? I'm not able to reproduce the error locally.

**Review by liyinan926 — APPROVED** [review]
LGTM.

**Comment by yuchaoran2011** [comment]
@liyinan926 @mrow4a I just verified my hypothesis that the Jenkins build machine running the `pull-charts-e2e` tests indeed had some uncleaned-up state, making my PR fail with "Error: namespaces "sparkoperator" already exists". In my latest commit, I simply updated the default namespace name from `sparkoperator` to `spark-operator` in `values.yaml`, then the build succeeded.

Now I think this PR is all set to be merged! If you need more time to try it out yourself, please do. When you are ready, Can you merge it? If you don't have the permissions, can you ping someone who does? Thanks!

**Comment by liyinan926** [comment]
@foxish Are you able to help merge this? I don't have the permission to merge. Thanks!

**Comment by yuchaoran2011** [comment]
@foxish Can you help merge it? Thanks!

File: incubator/sparkoperator/values.yaml
Comment (mrow4a) [inline incubator/sparkoperator/values.yaml:3]: You mention in a doc `sparkoperator`, but default is `spark-operator`. Change to `sparkoperator` ?

Comment (yuchaoran2011) [reply incubator/sparkoperator/values.yaml:3]: Good catch! I updated the doc instead because of the build issue I mentioned in an earlier comment.

File: incubator/sparkoperator/templates/spark-operator.yaml
Comment (mrow4a) [inline incubator/sparkoperator/templates/spark-operator.yaml]: It is super confusing with this having the same name as an `sparkoperator` itself. Since it will appear as a pod, which COMPLETED, and then actual spark operator will appear. Please rename it to something like `sparkoperator-init-job`, as you have `sparkoperator-cleanup-job`

File: incubator/sparkoperator/templates/spark-operator.yaml
Comment (mrow4a) [inline incubator/sparkoperator/templates/spark-operator.yaml:129]: Strange, I get container-creating on spark operator, and error 
```
  Warning  FailedMount  1m (x9 over 3m)  kubelet, test-cluster-tnrgym2mzvyk-minion-0  MountVolume.SetUp failed for volume "webhook-certs" : secrets "spark-webhook-certs" not found
```

File: incubator/sparkoperator/templates/spark-operator.yaml
Comment (yuchaoran2011) [reply incubator/sparkoperator/templates/spark-operator.yaml]: Makes sense. Done.

File: incubator/sparkoperator/templates/spark-operator.yaml
Comment (mrow4a) [reply incubator/sparkoperator/templates/spark-operator.yaml:129]: Hmm, interesting looking at gen-certs job output
```
HTTP/1.1 403 Forbidden
Content-Type: application/json
X-Content-Type-Options: nosniff
Date: Wed, 22 Aug 2018 13:59:47 GMT
Content-Length: 331
Connection: close

{
  "kind": "Status",
  "apiVersion": "v1",
  "metadata": {

  },
  "status": "Failure",
  "message": "secrets is forbidden: User \"system:serviceaccount:spark-operator:sparkoperator\" cannot create secrets in the namespace \"spark-operator\"",
  "reason": "Forbidden",
  "details": {
    "kind": "secrets"
  },
  "code": 403
}
```

Comment (yuchaoran2011) [reply incubator/sparkoperator/templates/spark-operator.yaml:129]: Weird. Working for me. I'll take a look today.

File: incubator/sparkoperator/templates/spark-operator-rbac.yaml
Comment (mrow4a) [inline incubator/sparkoperator/templates/spark-operator-rbac.yaml]: I will experiment adding here secrets, not sure if this helps

**Comment by liyinan926** [comment]
@yuchaoran2011 @mrow4a I'm thinking that we might want to put this chart into the operator repo instead for a few reasons: 1) it will have a much faster turn around time for PRs, 2) it makes users better be aware of the chart, and 3) it's much easier to keep the chart and the operator in sync. Thoughts?

**Comment by mrow4a** [comment]
Does `helm delete` cleans up everything properly for you? including `spark-operator` namespace?

**Comment by yuchaoran2011** [comment]
@liyinan926 Yeah the PR turnaround time on this repo is a PITA. Initially I thought what we should do is: get the PR merged here, then update the operator doc's installation section to point to this chart. The sync issue can be avoided by removing the `manifest` directory in the operator repo. All future changes should be made to the chart. One of the benefits of keeping the chart in this repo is to allow easy access because it's available in the default helm repo (the user doesn't need to run `helm add repo` or manually build the tgz).

But now it looks like when future changes/updates are required, it would take ages before they can be merged here. So I'm cool moving it to the operator repo.

**Comment by yuchaoran2011** [comment]
@mrow4a Yes. `helm delete` cleans up everything installed during `helm install` for me. Did you see uncleaned-up stuff?

File: incubator/sparkoperator/templates/spark-operator.yaml
Comment (mrow4a) [inline incubator/sparkoperator/templates/spark-operator.yaml]: I can see that cleanup job runs with a default namespace, but init job runs in a namespace. I think I encounter the same problem with the namespace deletion as with secret creation. Investigating now

Comment (yuchaoran2011) [reply incubator/sparkoperator/templates/spark-operator.yaml]: I put the cleanup job to run in the default namespace because I don't want the job to run in the namespace that it's trying to delete. The `"helm.sh/hook-delete-policy": hook-succeeded` ensures that the cleanup job itself is cleaned up after it's done. But I haven't been to reproduce errors you saw. I wonder if it's something that I have in my cluster but I forgot..

File: incubator/sparkoperator/templates/spark-operator.yaml
Comment (mrow4a) [reply incubator/sparkoperator/templates/spark-operator.yaml:129]: I debug the issue, and service accounts are created, but roles and clusterroles not for some strange reason. But helm tells me that it is created.. maybe some issue with cluster

Comment (mrow4a) [reply incubator/sparkoperator/templates/spark-operator.yaml:129]: Ok, they were created properly. Not sure what could be wrong.

File: incubator/sparkoperator/templates/spark-operator-rbac.yaml
Comment (mrow4a) [reply incubator/sparkoperator/templates/spark-operator-rbac.yaml]: @yuchaoran2011 maybe I am mistaken, but we are creating with `sparkoperator` token a a secret, but here we don't allow creating secrets? How is the batch job able to create it ? (adding it anyways did not solve my issue, but looking for a hint)

File: incubator/sparkoperator/templates/spark-operator.yaml
Comment (mrow4a) [reply incubator/sparkoperator/templates/spark-operator.yaml:129]: Interesting, running this batch job separately succeeded and I get spark operator running

Comment (yuchaoran2011) [reply incubator/sparkoperator/templates/spark-operator.yaml:129]: Hmm it might be a Helm issue. Now I remember running `kubectl create clusterrolebinding bind-cluster-admin --clusterrole=cluster-admin --serviceaccount=kube-system:default
`. Can you try that too?

**Comment by mrow4a** [comment]
@yuchaoran2011 is it possible that I get batch job executed before the `clusterrole` and clusterrolebinding` for `sparkoperator` exist? I can see that `serviceaccount` is created, but seems binding is not applied before batch job executes. Any clue?

File: incubator/sparkoperator/templates/spark-operator.yaml
Comment (mrow4a) [reply incubator/sparkoperator/templates/spark-operator.yaml]: I run the cleanup job separately, and I get this
```
HTTP/1.1 403 Forbidden
Content-Type: application/json
X-Content-Type-Options: nosniff
Date: Wed, 22 Aug 2018 16:11:36 GMT
Content-Length: 376

{
  "kind": "Status",
  "apiVersion": "v1",
  "metadata": {

  },
  "status": "Failure",
  "message": "namespaces \"spark-operator\" is forbidden: User \"system:serviceaccount:default:default\" cannot delete namespaces in the namespace \"spark-operator\"",
  "reason": "Forbidden",
  "details": {
    "name": "spark-operator",
    "kind": "namespaces"
  },
  "code": 403
}
```

Comment (mrow4a) [reply incubator/sparkoperator/templates/spark-operator.yaml]: Doing below made the cleanup-job work.. but not sure if this is a solution -@yuchaoran2011 @liyinan926 
```
kubectl create clusterrolebinding bind-cluster-admin-default --clusterrole=cluster-admin --serviceaccount=default:default
```

File: incubator/sparkoperator/templates/spark-operator.yaml
Comment (mrow4a) [reply incubator/sparkoperator/templates/spark-operator.yaml:129]: @yuchaoran2011 yes, I will do one more experiment tomorrow on trying to force clusterrole and binding to be before an init job.

**Comment by liyinan926** [comment]
> But now it looks like when future changes/updates are required, it would take ages before they can be merged here. So I'm cool moving it to the operator repo.

Cool. Let's wait for another few days and see this someone can help merge this. If not, we may just move it to the operator repo instead.

**Comment by liyinan926** [comment]
@viglesiasce can you help with reviewing and merging this chart?

File: incubator/sparkoperator/templates/spark-operator-rbac.yaml
Comment (mrow4a) [reply incubator/sparkoperator/templates/spark-operator-rbac.yaml]: @yuchaoran2011 you HAVE to add here `secrets` resources for this clusterrole

File: incubator/sparkoperator/templates/spark-operator.yaml
Comment (mrow4a) [inline incubator/sparkoperator/templates/spark-operator.yaml]: @yuchaoran2011 Remove this hook, as we don't need it if `sparkoperator-init` job is executed in `post-install`

File: incubator/sparkoperator/templates/spark-operator.yaml
Comment (mrow4a) [inline incubator/sparkoperator/templates/spark-operator.yaml]: @yuchaoran2011 Remove also here a hook

File: incubator/sparkoperator/templates/spark-operator.yaml
Comment (mrow4a) [inline incubator/sparkoperator/templates/spark-operator.yaml]: @yuchaoran2011 change to `post-install`, so certificates will be added as last (spark operator will wait on ContainerCreate until this is executed)

File: incubator/sparkoperator/templates/spark-operator.yaml
Comment (mrow4a) [inline incubator/sparkoperator/templates/spark-operator.yaml]: @yuchaoran2011 remove this job as it is not needed when namespace is not realised with hook

**Review by mrow4a — CHANGES_REQUESTED** [review]
In order to avoid "delete-resource.sh", and all the creation of namespace/serviceaccount with hooks, please run `sparkoperator-init` job as post-install. Additionaly, add to service account `sparkoperator` permission to create secrets (https://github.com/helm/charts/pull/6239#discussion_r212230814). I have given related comments above. You can find full diff in https://stikked.web.cern.ch/stikked/view/18a54853.

With above comments addressed, it all works, LGTM

File: incubator/sparkoperator/templates/spark-operator.yaml
Comment (mrow4a) [reply incubator/sparkoperator/templates/spark-operator.yaml]: Add also `"helm.sh/hook-delete-policy": hook-succeeded`

**Comment by yuchaoran2011** [comment]
@mrow4a Now your comments are incorporated. Using a `post-install` hook indeed simplified the chart. Thanks! I'll wait a few more days to see if @foxish is able to merge it. If not, I'll move this PR to the operator repo.

**Review by mrow4a — APPROVED** [review]
Tested and looks good

**Comment by liyinan926** [comment]
@foxish need lgtm and approval from you. Thanks!

**Pull request merged**
